core/unix/os.c

```diff
@@ -7834,13 +7834,28 @@ mmap_check_for_module_overlap(app_pc base, size_t size, bool readable, uint64 in
                  * Note, if it is a header of a different module, then we'll not have
                  * an overlap, so we will not hit this case.
                  */
-                ASSERT_CURIOSITY(
-                    ma->start + ma->os_data.alignment ==
-                    base
-                        /* On Mac we walk the dyld module list before the
-                         * address space, so we often hit modules we already
-                         * know about. */
-                        IF_MACOS(|| !dynamo_initialized && ma->start == base));
+#ifdef AARCH64
+                /* For AArch64 systems with 64K pages it is possible for small modules
+                 * (less than 1 page of disk space) to have multiple RW LOADable data
+                 * segments. Since these segments are mapped from a single disk space they
+                 * will all have a elf_header satisfying the check above.
+                 */
+                if (ma->os_data.alignment == 65536) {
+                    size_t offset = base - ma->start;
+                    size_t seg_id = offset / ma->os_data.alignment;
+                    ASSERT_CURIOSITY(seg_id < ma->os_data.num_segments &&
+                                     ma->os_data.segments[seg_id].start == base &&
+                                     ma->os_data.segments[seg_id].prot ==
```

Comment:
I think what we want is to walk the segment array until we hit one whose start==base and check that `module_segment_data_t.offset < PAGE_SIZE` for every one through that one.

Reply:
Yep, that's a better check and what I should have done in the first place. For traversing the array maybe we can avoid it  based on the assumption that the alignment between the segments in the segment array seems to be constant. If my assumption is wrong or checking all the predecessors of the current segment is necessary (the predecessors will be checked on their own when they are mapped to memory and I would think that checking the offset of the current segment would suffice), I will switch to just traversing and checking all of them.

Reply:
Given the scenario of non-adjustment segment pointed it out in another comment, it is better to do the traversal instead of computing the index of the segment directly from the base address.